[Model] Whisper model implementation

aurickq opened the pull request:
Add Whisper model implementation. Based on https://github.com/vllm-project/vllm/pull/5964 but heavily optimized and integrated with newer encoder/decoder support in vLLM. Currently only supports audio up to 30s. No VAD model, no beam search, no phoneme model, forced alignment, etc.

Performance overall looks OK, especially at larger batch sizes

<img width="679" alt="Screenshot 2024-12-18 at 5 33 15 PM" src="https://github.com/user-attachments/assets/7f91f018-8c2a-4040-9f59-b5f92bc59cd9" />

FIX https://github.com/vllm-project/vllm/issues/180

sfc-gh-aqiao committed:
add model def

sfc-gh-aqiao committed:
WIP

sfc-gh-aqiao committed:
WIP, passes profile run

aurickq committed:
Merge remote-tracking branch 'upstream/main' into whisper

sfc-gh-aqiao committed:
WIP

sfc-gh-aqiao committed:
kinda working with encoder decoder

sfc-gh-aqiao committed:
add whisper example

sfc-gh-aqiao committed:
update

sfc-gh-aqiao committed:
cleanup a bit

sfc-gh-aqiao committed:
batching

sfc-gh-aqiao committed:
flash_attn

sfc-gh-aqiao committed:
WIP (broken)

sfc-gh-aqiao committed:
WIP

sfc-gh-aqiao committed:
13rps

sfc-gh-aqiao committed:
fuse qkv

sfc-gh-aqiao committed:
clean

sfc-gh-aqiao committed:
20 RPS

sfc-gh-aqiao committed:
26rps

sfc-gh-aqiao committed:
41 rps

sfc-gh-aqiao committed:
fix tokenizer

sfc-gh-aqiao committed:
fix tp

sfc-gh-aqiao committed:
clean

sfc-gh-aqiao committed:
clean

sfc-gh-aqiao committed:
udpate

sfc-gh-aqiao committed:
add test

DarkLight1337 commented on vllm/worker/enc_dec_model_runner.py:
```diff
@@ -280,19 +280,19 @@ def profile_run(self) -> None:
         for group_id in range(max_num_seqs):
             seq_len = (max_num_batched_tokens // max_num_seqs +
                        (group_id < max_num_batched_tokens % max_num_seqs))
+            seq_len = min(seq_len, self.model_config.max_model_len)
```
Hmm, I think `seq_len` should never be calculated to be greater than `max_model_len` in the first place.

DarkLight1337 commented on vllm/transformers_utils/tokenizer_group/tokenizer_group.py:
```diff
@@ -57,7 +59,11 @@ def encode(self,
                request_id: Optional[str] = None,
                lora_request: Optional[LoRARequest] = None) -> List[int]:
```
To be consistent with HF, I think `add_special_tokens` should be an argument to `encode` and `encoder_async` instead of the initializer.

DarkLight1337 commented on vllm/model_executor/models/registry.py:
```diff
@@ -165,6 +165,7 @@
     "UltravoxModel": ("ultravox", "UltravoxModel"),
     # [Encoder-decoder]
     "MllamaForConditionalGeneration": ("mllama", "MllamaForConditionalGeneration"),  # noqa: E501
+    "WhisperForConditionalGeneration": ("whisper", "WhisperForConditionalGeneration"),  # noqa: E501
```
Please update the Supported Models page and `tests/models/registry.py` accordingly.

DarkLight1337 commented on vllm/inputs/preprocess.py:
```diff
@@ -439,8 +439,15 @@ def _build_enc_dec_llm_inputs(
         if decoder_inputs is None:
-            dec_token_ids = self._prepare_decoder_input_ids_for_generation(
-                None)
+            if self.model_config.hf_config.model_type == "whisper":
+                # For Whisper models, the text prompt should go to the decoder.
+                # If no explicit encoder/decoder inputs, then copy the prompt
+                # from the encoder to the decoder. The encoder tokens are later
+                # overridden by the audio features.
+                dec_token_ids = encoder_inputs["prompt_token_ids"].copy()
+            else:
+                dec_token_ids = self._prepare_decoder_input_ids_for_generation(
+                    None)
```
Is there a way to determine this without model type information?

aurickq replied:
I am not sure about generalizing this from a single example. In the long term it may be better to allow the model definition to specify exactly the mapping between input fields and where they go (e.g. encoder/decoder)

afeldman-nm replied:
Agreed with @aurickq , long-term it is probably best to either (1) have the model definition specify whether to map the input text prompt to the encoder, or (2) add a default behavior only for *multi-modal models with cross-attention*, wherein the text prompt is always routed to the decoder & the non-text modality is always mapped to the encoder.

(I worked on adding encoder/decoder cross-attention support to v0)

DarkLight1337 commented on vllm/model_executor/models/whisper.py:
```diff
@@ -0,0 +1,689 @@
+
+
+@INPUT_REGISTRY.register_dummy_encoder_data(dummy_encoder_data_for_whisper)
+@INPUT_REGISTRY.register_input_processor(input_processor_for_whisper)
+@MULTIMODAL_REGISTRY.register_input_mapper("audio", input_mapper_for_whisper)
```
If possible, use the merged multi-modal processor (#10114) which should simplify your code.

aurickq replied:
What's the right approach here? I tried the Processor and there appears to be several mismatched assumptions with Whisper and encoder-decoder models:

1. There is a mandatory `_get_prompt_replacements` which does not apply to Whisper (audio tokens never get embedded within text tokens).
2. The processor assumes one token sequence but for Whisper, two token sequences are needed: one as a placeholder for the encoder to allocate the correct cross-attention kv-cache, and the second for the actual prompt which only goes to the decoder.

(1) is easy to get around with some awkwardness, but (2) seems only supported by the legacy multimodal processing pipeline which returns two different inputs, one for the encoder and another for the decoder.

DarkLight1337 replied:
Thanks for bringing this up! I haven't thought about encoder-decoder models for the new pipeline yet... let's use the legacy processing pipeline for now then.

DarkLight1337 reviewed (commented):
Some initial comments.

DarkLight1337 commented on vllm/model_executor/models/whisper.py:
```diff
@@ -0,0 +1,689 @@
+logger = init_logger(__name__)
+
+
+def sinusoids(
+        length: int, channels: int, max_timescale: float = 10000
+) -> torch.Tensor:
+    """Returns sinusoids for positional embedding"""
+    if channels % 2 != 0:
+        raise ValueError(
+            f"Number of channels has to be divisible by 2 for sinusoidal "
+            f"positional embeddings, got {channels} channels."
+        )
+    log_timescale_increment = math.log(max_timescale) / (channels // 2 - 1)
+    inv_timescales = torch.exp(-log_timescale_increment *
+                               torch.arange(channels // 2))
+    scaled_time = torch.arange(length).view(-1, 1) * inv_timescales.view(1, -1)
+    return torch.cat([scaled_time.sin(), scaled_time.cos()], dim=1)
```
Let's import from `transformers` if the implementation is exactly the same as theirs.

DarkLight1337 commented on vllm/model_executor/models/whisper.py:
```diff
@@ -0,0 +1,689 @@
+            quant_config=quant_config,
+            prefix=f"{prefix}.self_attn",
+        )
+        self.activation_fn = FastGELU()
```
Do we not support other activation functions?

aurickq replied:
use `config.activation_function` now

DarkLight1337 commented on vllm/model_executor/models/whisper.py:
```diff
@@ -0,0 +1,689 @@
+                                                config.vocab_size, logit_scale)
+        self.sampler = Sampler()
+
+    def forward(
```
To support V1, the model needs to implement `get_input_embeddings` and `get_multimodal_embeddings`.

aurickq replied:
Done, but there is some awkwardness
- `get_multimodal_embeddings` for Whisper also requires kv-cache and attn metadata which is not in the base method's interface
- It looks like `get_input_embeddings` expects to embed multimodal tokens into text tokens. again, this is not what Whisper does. Currently I just have this method return the decoder token embeddings.

I couldn't find anywhere in vLLM that uses these two methods so it's hard for me to understand their intended usage. Happy to change the implementation to whatever works better.

DarkLight1337 replied:
Sorry, I forgot that V1 doesn't support encoder decoder yet. Let's just add a TODO and come back to this later.

aurickq replied:
added todos

Isotr0py reviewed (commented):
Just some initial comments about model implementation :)

sfc-gh-aqiao committed:
some cleanup

sfc-gh-aqiao committed:
formatting

sfc-gh-aqiao committed:
format

sfc-gh-aqiao committed:
mypy

sfc-gh-aqiao committed:
mypy

sfc-gh-aqiao committed:
format

mgoin commented on requirements-common.txt:
```diff
@@ -35,3 +35,4 @@ setuptools>=74.1.1; python_version > '3.11' # Setuptools is used by triton, we n
 einops # Required for Qwen2-VL.
 compressed-tensors == 0.8.0 # required for compressed-tensors
 depyf==0.18.0 # required for profiling and debugging torch.compile
+librosa >= 0.10.2  # required for audio processing including Whisper
```
I believe we made librosa optional in the past due to licensing concerns, I can find the issue that started this if needed

We currently have it as a dep through the vllm[audio] extra here https://github.com/vllm-project/vllm/blob/fdea8ec16775e1645620b5ff46b799d60df4624c/setup.py#L633

mgoin commented on vllm/model_executor/models/whisper.py:
```diff
@@ -0,0 +1,702 @@
+from functools import lru_cache
+from typing import Iterable, List, Mapping, Optional, Set, Tuple, Union
+
+import librosa
```
Related to librosa in the audio extra, we should make this a lazy import

sfc-gh-aqiao committed:
fix tests

DarkLight1337 commented on vllm/transformers_utils/tokenizer_group/tokenizer_group.py:
```diff
@@ -55,19 +55,31 @@ def _raise_if_input_too_long(self,
         tokenizer = self.get_lora_tokenizer(lora_request)
-        ret = tokenizer.encode(prompt)
+        if add_special_tokens is not None:
+            ret = tokenizer.encode(  # type: ignore[call-arg]
+                prompt,
+                add_special_tokens=add_special_tokens)
```
Let's move `vllm.multmodal.processor._encode` to `vllm.transformers_utils` so we can reuse the logic and avoid this type error.

sfc-gh-aqiao committed:
librosa

sfc-gh-aqiao committed:
Merge remote-tracking branch 'vllm-project/main' into whisper

sfc-gh-aqiao committed:
small

Isotr0py commented on vllm/model_executor/models/whisper.py:
```diff
@@ -0,0 +1,711 @@
+    )
+
+
+def _resample(audio: np.ndarray, orig_sr: int, target_sr: int) -> np.ndarray:
+    try:
+        import librosa
+    except ImportError as exc:
+        raise ImportError(
+            "Please install vllm[audio] for audio support.") from exc
+    return librosa.resample(audio, orig_sr=orig_sr, target_sr=target_sr)
```
Seems that this is totally same with the `resample_audio` in `vllm/multimodal/audio.py`?

Isotr0py commented on vllm/model_executor/models/whisper.py:
```diff
@@ -0,0 +1,711 @@
+    )
+
+
+@lru_cache
+def get_whisper_processor(
+    processor_name: str,
+    *args,
+    trust_remote_code: bool = False,
+    revision: Optional[str] = None,
+    **kwargs,
+) -> WhisperProcessor:
+    """Gets an whisper processor for the given model name via HuggingFace."""
+    return WhisperProcessor.from_pretrained(
+        processor_name,
+        *args,
+        trust_remote_code=trust_remote_code,
+        revision=revision,
+        **kwargs,
+    )
```
Perhaps we can use `cached_get_processor` from `vllm/transformers_utils/processor.py`?

sfc-gh-aqiao committed:
updates

sfc-gh-aqiao committed:
lint

sfc-gh-aqiao committed:
add todos

sfc-gh-aqiao committed:
bugfix

orgh0 commented:
Hey @sfc-gh-aqiao  , thanks so much for taking this forward. It's going to help a lot of folks. 

Quick question:- By when do you estimate this feature would get merged? 
trying to understand if i could start using this branch for the work we do. Given it's close to merging.

sfc-gh-aqiao committed:
Merge remote-tracking branch 'vllm-project/main' into whisper

sfc-gh-aqiao committed:
fix repeating issue

sfc-gh-aqiao committed:
fix tests

aurickq commented:
> Hey @sfc-gh-aqiao , thanks so much for taking this forward. It's going to help a lot of folks.
> 
> Quick question:- By when do you estimate this feature would get merged? trying to understand if i could start using this branch for the work we do. Given it's close to merging.

Just fixed a few minor issues and rebased. It should be ready to go, and it works well in my tests. Just waiting on approval or further reviews.

mgoin commented on vllm/model_executor/models/whisper.py:
```diff
@@ -0,0 +1,736 @@
+    num_tokens = ctx.model_config.hf_config.max_source_positions
+    return DummyData(
+        SequenceData.from_prompt_token_counts((0, num_tokens)),
+        {"audio": [(np.zeros(30 * sample_rate), sample_rate)]},
```
Is the 30 seconds here just hardcoded in the model, or is there some place we can read for this limit?

aurickq replied:
got rid of hard coding

mgoin commented on vllm/model_executor/models/whisper.py:
```diff
@@ -0,0 +1,736 @@
+                                   mm_counts: Mapping[str, int]):
+    assert mm_counts["audio"] == 1
+    sample_rate = 16000
+    num_tokens = ctx.model_config.hf_config.max_source_positions
```
nit: 
```suggestion
    num_tokens = get_max_whisper_audio_tokens(ctx)
```

mgoin commented on vllm/model_executor/models/whisper.py:
```diff
@@ -0,0 +1,736 @@
+    audio = resample_audio(audio, orig_sr=orig_sr, target_sr=target_sr)
+    multi_modal_data["audio"] = (audio, target_sr)
+    # Pre-allocate placeholder tokens in encoder sequence
+    num_tokens = ctx.model_config.hf_config.max_source_positions
```
```suggestion
    num_tokens = get_max_whisper_audio_tokens(ctx)
```

mgoin commented on vllm/model_executor/models/whisper.py:
```diff
@@ -0,0 +1,736 @@
+                       sampling_rate=sampling_rate,
+                       return_tensors="pt")
+    kwargs["input_features"] = kwargs["input_features"].squeeze(0)
+    kwargs["input_features"] = kwargs["input_features"].to(torch.float16)
```
Do we support running a whisper model in bfloat16?

aurickq replied:
yes, got rid of this hard coding

mgoin reviewed (commented):
Thanks for the nice work! I think this should be ready to go by me - please add the test to the buildkite pipeline to verify

sfc-gh-aqiao committed:
clean

aurickq commented on tests/models/encoder_decoder/audio/test_whisper.py:
done, thanks

sfc-gh-aqiao committed:
reduce tests

aurickq commented on .buildkite/test-pipeline.yaml:
```diff
@@ -363,12 +363,14 @@ steps:
   - tests/models/decoder_only/audio_language
   - tests/models/decoder_only/vision_language
   - tests/models/embedding/vision_language
+  - tests/models/encoder_decoder/audio
```
done

aurickq commented on tests/models/encoder_decoder/audio/test_whisper.py:
```diff
@@ -0,0 +1,109 @@
+@fork_new_process_for_each_test
+@pytest.mark.core_model
+@pytest.mark.parametrize("model",
+                         ["openai/whisper-medium", "openai/whisper-large-v3"])
```
changed to `small` (~200M) and `large-v3-turbo` (~800M). the tiny model outputs are poor so not sure if it's worth testing. also nice to have a more recent model (`large-v3-turbo`)

sfc-gh-aqiao committed:
format

mgoin reviewed (approved):
LGTM and verified the example locally. Last part is adding it to the model docs, feel free to do in a followup. cc @DarkLight1337 @Isotr0py for final pass

aurickq committed:
Update offline_inference_whisper.py

DarkLight1337 commented on tests/models/registry.py:
```diff
@@ -128,6 +128,7 @@ class _HfExamplesInfo:
     # [Encoder-decoder]
     "BartModel": _HfExamplesInfo("facebook/bart-base"),
     "BartForConditionalGeneration": _HfExamplesInfo("facebook/bart-large-cnn"),
+    "WhisperForConditionalGeneration": _HfExamplesInfo("openai/whisper-large-v3"),  # noqa: E501
```
Let's move this under Mllama since it's a multimodal model.

DarkLight1337 commented on vllm/inputs/preprocess.py:
```diff
@@ -184,10 +184,16 @@ def _tokenize_prompt(
         """
         tokenizer = self.get_tokenizer_group()
-
+        add_special_tokens = None
+        if self.model_config.hf_config.model_type == "whisper":
+            # For Whisper, special tokens should be provided by the user based
+            # on the task and language of their request. Also needed to avoid
+            # appending an EOS token to the prompt which disrupts generation.
+            add_special_tokens = False
```
Can we remove the eos token in the input processor? But maybe we can improve this later via the merged multi-modal processor since I assume HF can handle this automatically.

Isotr0py replied:
Seems that `<|startoftranscript|>` is also a special token which will be added by tokenizer by default:
```python3
from transformers import AutoProcessor

processor = AutoProcessor.from_pretrained("openai/whisper-small")
tokenizer = processor.tokenizer
encode = tokenizer("<|startoftranscript|>")
print(tokenizer.decode(encode["input_ids"]))
```

It outputs `'<|startoftranscript|><|notimestamps|><|startoftranscript|><|endoftext|>'`, which means that we need to remove a `"<|startoftranscript|>"` to get the original prompt.

Perhaps we need to construct user prompts with non special tokens.

aurickq replied:
Yes, there are other special tokens that are added as well. There are also some other challenges
1. Some tokens are not reliably mapped to the same ids. E.g. `<|notimestamps|>` is 50363 for whisper-small but 50364 for whisper-large-v3. This makes it awkward to try to detect these special tokens from inside the input processor.
2. From in the input processor, not sure how to reliably differentiate between the two cases
  a. User passed in text `prompt` which is tokenized in vLLM. We need to strip special tokens.
  b. User passed in tokens `prompt_token_ids` which should not be stripped.

Basically, I'm not sure how to reliably remove special tokens after it's already encoded. It seems error-prone and brittle to third-party behavior in huggingface. I understand that having the tokenizer not add special tokens in the first place involves more changes inside vLLM, but IMO it's the most reliable solution that's more robust to future changes in vLLM and huggingface.

sfc-gh-aqiao committed:
move whisper test registry

DarkLight1337 reviewed (approved):
Let's merge this first then.

DarkLight1337 commented:
Can you merge from main to fix the CI failures?

sfc-gh-aqiao committed:
Merge remote-tracking branch 'vllm-project/main' into whisper

DarkLight1337 commented:
Thanks again for your hard work!

ArmykOliva commented:
Thank you for your work. Are Rest API endpoints reachable as well? Could you add an example for that please?

aryan1165 commented:
Just a quick question, Is anyone implementing VAD, to support chunking and accepting audios of len more than 30s. Also how continuous batching can be supported?

Temirulan commented:
Thanks for great news! Few questions here:

- [ ] Is word timestamping working using this upload?
- [ ] Does this support inflight batching?

aryan-sarvam commented:
Hey, I tried to use whisper in online way, using vllm.entrypoints.openai api server.py but, we can't hit Completions API (Can't send multimodal data), as mentioned in the Docs (generate and Completions API are similar), and chat_utils.py doesnt have whisper support. 

Is someone working on that part. I tried some rough solutions but couldn't get to execute properly. 

I will also open an issue regarding this, if this is not yet on the pipeline someone building.
Thanks

aryan-sarvam commented:
> Thank you for your work. Are Rest API endpoints reachable as well? Could you add an example for that please?

Hi, I dont think they are, Completions one is bound to support this but it doesn't have multi modal support, and chat doesn't have this support. Let me know if you came up with a solution?

aryan-sarvam commented:
> Add Whisper model implementation. Based on #5964 but heavily optimized and integrated with newer encoder/decoder support in vLLM. Currently only supports audio up to 30s. No VAD model, no beam search, no phoneme model, forced alignment, etc.
> 
> Performance overall looks OK, especially at larger batch sizes
> 
> <img alt="Screenshot 2024-12-18 at 5 33 15 PM" width="679" src="https://private-user-images.githubusercontent.com/6137920/397146758-7f91f018-8c2a-4040-9f59-b5f92bc59cd9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0ODU3MDAsIm5iZiI6MTczNjQ4NTQwMCwicGF0aCI6Ii82MTM3OTIwLzM5NzE0Njc1OC03ZjkxZjAxOC04YzJhLTQwNDAtOWY1OS1iNWY5MmJjNTljZDkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDExMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMTBUMDUwMzIwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmQwMzljOTI1OTI3ZjQ2YmZkYzM2NDhiNDg5MDAyYzg4YTBlYjVmM2YxYzNkMGY1M2EwMmYxNzhlNzhhMDEyMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.BNO2k2TToRP5yY_dy6SZYpcbqDtOb9moMGt7rcf4mXM">
> FIX #180

Hey, I was doing benchmarking on H100, of audio len around 18s. I found out that RTFx maxes out at a really low value around 120 at batch size 96. Can you share the benchmarking parameters or flags so that we can reproduce these results.

aurickq commented:
> > Add Whisper model implementation. Based on #5964 but heavily optimized and integrated with newer encoder/decoder support in vLLM. Currently only supports audio up to 30s. No VAD model, no beam search, no phoneme model, forced alignment, etc.
> > Performance overall looks OK, especially at larger batch sizes
> > <img alt="Screenshot 2024-12-18 at 5 33 15 PM" width="679" src="https://private-user-images.githubusercontent.com/6137920/397146758-7f91f018-8c2a-4040-9f59-b5f92bc59cd9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0ODU3MDAsIm5iZiI6MTczNjQ4NTQwMCwicGF0aCI6Ii82MTM3OTIwLzM5NzE0Njc1OC03ZjkxZjAxOC04YzJhLTQwNDAtOWY1OS1iNWY5MmJjNTljZDkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDExMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMTBUMDUwMzIwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmQwMzljOTI1OTI3ZjQ2YmZkYzM2NDhiNDg5MDAyYzg4YTBlYjVmM2YxYzNkMGY1M2EwMmYxNzhlNzhhMDEyMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.BNO2k2TToRP5yY_dy6SZYpcbqDtOb9moMGt7rcf4mXM">
> > FIX #180
> 
> Hey, I was doing benchmarking on H100, of audio len around 18s. I found out that RTFx maxes out at a really low value around 120 at batch size 96. Can you share the benchmarking parameters or flags so that we can reproduce these results.

Hi, I am running the example `python examples/offline_inference/whisper.py` which measures the time to run ~2K audio clips:
```
Duration: 57.445557594299316
RPS: 35.65114668158841
```

aryan-sarvam commented:
> Hi, I am running the example `python examples/offline_inference/whisper.py` which measures the time to run ~2K audio clips:

Hi, I ran the same exact script. The results were -

```Duration: 325.3493559360504
RPS: 6.294771950931872
```

Are there any other flags i have to use to reproduce the results?

aryan-sarvam commented:
> Duration: 57.445557594299316
> RPS: 35.65114668158841

Hi @aurickq , I build the project from source and ran the same file. On h100, with exact same file, results are :-

```Duration: 309.14796566963196
RPS: 6.624659475160758
````

Can you tell if there are some other things to do? If possible Can you connect with me, sometime for proper setup. 

Thanks.

mru4913 commented:
@aurickq @DarkLight1337 is `docker run` ready for this model yet?

DarkLight1337 commented:
No, please see #12909

mru4913 commented:
@aryan-sarvam do you have any update ?
